hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/MultipleSparkJobExecutionStrategy.java

```diff
@@ -420,64 +449,131 @@ writeConfig, new StoragePath(bootstrapFilePath)), partitionFields,
+
+      RDD<InternalRow> internalRowRDD = jsc.parallelize(clusteringOps, clusteringOps.size()).flatMap(new FlatMapFunction<ClusteringOperation, InternalRow>() {
+        @Override
+        public Iterator<InternalRow> call(ClusteringOperation clusteringOperation) throws Exception {
+          // construct FileSlice to pass into FileGroupReader
+          String partitionPath = clusteringOperation.getPartitionPath();
+          HoodieBaseFile baseFile = new HoodieBaseFile(new StoragePath(basePath, clusteringOperation.getDataFilePath()).toString());
+          List<HoodieLogFile> logFiles = clusteringOperation.getDeltaFilePaths().stream().map(p ->
+                  new HoodieLogFile(new StoragePath(FSUtils.constructAbsolutePath(
+                      basePath, partitionPath), p)))
+              .collect(Collectors.toList());
+          FileSlice fileSlice = new FileSlice(new HoodieFileGroupId(partitionPath, clusteringOperation.getFileId()), baseFile.getCommitTime(), baseFile, logFiles);
+
+          // instantiate other supporting cast
+          Schema readerSchema = serializableTableSchemaWithMetaFields.get();
+          Option<InternalSchema> internalSchemaOption = Option.empty();
+          if (isInternalSchemaPresent) {
+            internalSchemaOption = SerDeHelper.fromJson(internalSchemaStr);
           }
-          return readPaths;
-        })
-        .flatMap(Collection::stream)
-        .filter(path -> !path.isEmpty())
-        .map(StoragePath::new)
-        .toArray(StoragePath[]::new);
-
-    HashMap<String, String> params = new HashMap<>();
-    if (hasLogFiles) {
-      params.put("hoodie.datasource.query.type", "snapshot");
-    } else {
-      params.put("hoodie.datasource.query.type", "read_optimized");
-    }
+          Option<HoodieReaderContext> readerContextOpt = broadcastManager.retrieveFileGroupReaderContext(new StoragePath(basePath));
+          Configuration conf = broadcastManager.retrieveStorageConfig().get();
+          HoodieTableMetaClient localMetaClient = HoodieTableMetaClient.builder()
+              .setBasePath(basePath)
+              .setConf(new HadoopStorageConfiguration(conf))
+              .build();
 
-    StoragePath[] paths;
-    if (hasLogFiles) {
-      String rawFractionConfig = getWriteConfig().getString(HoodieMemoryConfig.MAX_MEMORY_FRACTION_FOR_COMPACTION);
-      String compactionFractor = rawFractionConfig != null
-          ? rawFractionConfig : HoodieMemoryConfig.DEFAULT_MR_COMPACTION_MEMORY_FRACTION;
-      params.put(HoodieMemoryConfig.MAX_MEMORY_FRACTION_FOR_COMPACTION.key(), compactionFractor);
+          // instantiate FG reader
+          HoodieFileGroupReader<T> fileGroupReader = new HoodieFileGroupReader<>(
+              readerContextOpt.get(),
+              localMetaClient.getStorage().newInstance(new StoragePath(basePath), new HadoopStorageConfiguration(conf)),
```

Comment:
This is unnecessary now. You can directly use `localMetaClient` which contains the correct `Configuration` instance.